receiver.go

```diff
@@ -60,123 +69,154 @@ func (r *receiver) receiveLoop() error {
-		return nil
+	if r.context.LogFrames {
+		r.context.logFrame("Received frame: %s (flags=%8b, length=%d)",
+			frameString(requestNumber, flags), flags, r.frameBuffer.Len())
 	}
 
-	frameSize, err := writeFull(frame, msgStream.writer)
+	r.frameDecoder.enableCompression(compressed)
+	frame, err = r.frameDecoder.readAll()
 	if err != nil {
+		r.context.log("Error decompressing frame %s: %v", frameString(requestNumber, flags), err)
 		return err
 	}
-	if complete {
-		msgStream.writer.Close()
-	} else {
-		//FIX: This isn't the right place to do this, because this goroutine doesn't block even
-		// if the client can't read the message fast enough. The right place to send the ACK is
-		// in the goroutine that's running msgStream.writer. (Somehow...)
-		oldWritten := msgStream.bytesWritten
-		msgStream.bytesWritten += uint64(frameSize)
-		if oldWritten > 0 && (oldWritten/kAckInterval) < (msgStream.bytesWritten/kAckInterval) {
-			r.sender.sendAck(requestNumber, msgType, msgStream.bytesWritten)
+
+	if isACK {
+		bytesReceived, n := binary.Uvarint(frame)
+		if n > 0 {
+			r.sender.receivedAck(requestNumber, msgType-4, bytesReceived)
+		} else {
+			r.context.log("Error reading ACK frame")
 		}
+		return nil
+	}
 
+	curChecksum := r.frameDecoder.getChecksum()
+	if curChecksum != checksum {
+		return fmt.Errorf("Frame %s has invalid checksum %x; should be %x",
+			frameString(requestNumber, flags), curChecksum, checksum)
+	}
+
+	// Look up or create the writer stream for this message:
+	complete := (flags & kMoreComing) == 0
+	var msgStream *msgStreamer
+	if msgStream, err = r.getMessageStream(requestNumber, flags, frame, complete); err != nil {
+		return err
+	}
+
+	// Write the decoded frame body to the stream:
+	if msgStream != nil {
+		if frameSize, err := writeFull(frame, msgStream.writer); err != nil {
+			return err
+		} else if complete {
+			msgStream.writer.Close()
+		} else {
+			//FIX: This isn't the right place to do this, because this goroutine doesn't block even
+			// if the client can't read the message fast enough. The right place to send the ACK is
+			// in the goroutine that's running msgStream.writer. (Somehow...)
+			oldWritten := msgStream.bytesWritten
+			msgStream.bytesWritten += uint64(frameSize)
+			if oldWritten > 0 && (oldWritten/kAckInterval) < (msgStream.bytesWritten/kAckInterval) {
+				r.sender.sendAck(requestNumber, msgType, msgStream.bytesWritten)
+			}
+		}
 	}
 	return nil
 }
 
-// Why isn't this in the io package already, when ReadFull is?
-func writeFull(buf []byte, writer io.Writer) (nWritten int, err error) {
-	for len(buf) > 0 {
-		var n int
-		n, err = writer.Write(buf)
-		if err != nil {
-			break
+func (r *receiver) getMessageStream(requestNumber MessageNumber, flags frameFlags, frame []byte, complete bool) (*msgStreamer, error) {
+	var msgStream *msgStreamer
+	msgType := MessageType(flags & kTypeMask)
+	switch msgType {
+	case RequestType:
+		msgStream = r.pendingRequests[requestNumber]
```

Comment:
This looks like it might need to be protected by:

```
r.pendingMutex.Lock()
defer r.mutex.Unlock()	
```

Since it seems possible to be concurrently accessed by `backlog()` below.